idrstream/CP_idr.py

```diff
@@ -0,0 +1,391 @@
+        ----------
+        aspera_path : pathlib.Path
+            path to aspera sdk
+        aspera_key_path : pathlib.Path
+            path to aspera ssh key
+        screens_path : pathlib.Path
+            path to screens file used to locate IDR data
+        """
+        self.downloader = download.AsperaDownloader(
+            aspera_path, aspera_key_path, screens_path, self.idr_id
+        )
+        self.logger.info("Aspera downloader initialized")
+
+    def init_preprocessor(self, fiji_path: pathlib.Path):
+        """
+        initialize basicpy preprocessor
+
+        Parameters
+        ----------
+        fiji_path : pathlib.Path
+            path to Fiji.app folder
+        """
+        self.preprocessor = preprocess.BasicpyPreprocessor(fiji_path)
+        self.logger.info("Basicpy preprocessor initialized")
+
+    def convert_tsv_to_csv(
+        self, data_to_process_tsv: pathlib.Path, metadata_save_path: pathlib.Path
+    ):
+        """
+        Convert `data_to_process.tsv` file into a .csv for CellProfiler to read in when processing the images
+
+        Parameters
+        ----------
+        data_to_process_tsv : str
+            string path to the `data_to_process.tsv` file
+        """
+        # read in metadata tsv file
+        data_to_process_csv = pd.read_csv(data_to_process_tsv, sep="\t")
+
+        # converting tsv file into csv
+        data_to_process_csv.to_csv(metadata_save_path, index=False)
+
+        self.logger.info("CellProfiler metadata has been converted and saved")
+
+    def prepare_batch(self, batch_metadata: pd.DataFrame):
+        """
+        download data from a batch image and run further processes
+        image data is saved in the CP project folder
+
+        Parameters
+        ----------
+        batch_metadata : pd.DataFrame
+            metadata of images to extract features from
+        """
+        # iterate through image metadatas and download, preprocess, and segment each frame into the batch project
+        for index, row in batch_metadata.iterrows():
+            plate = row["Plate"]
+            well = row["Well"]
+            well_num = row["Well Number"]
+            # load frames saved as "x,y,z" and convert to [x,y,z]
+            frame_nums = str(row["Frames"]).split(",")
+            frame_nums = [int(frame_num) for frame_num in frame_nums]
+            self.logger.info(
+                f"Processing well {well} from {plate}, frames {frame_nums}"
+            )
+
+            download_save_path = pathlib.Path(f"{self.tmp_dir}/downloads/{plate}/")
+            well_movie_path = self.downloader.download_image(
+                plate, well_num, download_save_path
+            )
+            self.logger.info(f"well_movie_path: {well_movie_path}")
+
+            # give time for movie to fully save before trying to open it
+            # otherwise ImageJ tries to open to movie before it has been completely saved and it errors out
+            time.sleep(0.3)
```

Comment:
I can see why the `sleep` may be needed here but wonder why the movie may not be fully saved before a system call returns. In general, this could yield uncertain results depending on the system (I imagine processing time could vary). Do we know more about why the movie may not be fully saved in time for the subsequent lines to run without `sleep`? If resolving this via code would be outside the span of this PR (I trust your judgement here), It might make sense to document things more thoroughly within the comments.

Reply:
Unfortunately, I do not know why the movie may not be fully saved in time for the subsequent lines to run. I think it would be most accurate to add something like:

"Aspera downloader called in self.downloader.download_image can take extra time to finalize download depending on network speed"

Reply:
:+1:  Updated the wording